pyccel/parser/semantic.py

```diff
@@ -362,12 +362,19 @@ def insert_class(self, cls, parent=False):
         else:
             raise TypeError('Expected A class definition ')
 
+    def insert_template(self, expr):
+        """insert a template"""
+        self.namespace.templates.append(expr)
+
     def insert_header(self, expr):
         """."""
         if isinstance(expr, MethodHeader):
             self.namespace.headers[expr.name] = expr
         elif isinstance(expr, FunctionHeader):
-            self.namespace.headers[expr.func] = expr
+            if expr.func in self.namespace.headers:
+                self.namespace.headers[expr.func].append(expr)
+            else:
+                self.namespace.headers[expr.func] = [expr]
```

Comment:
Is there a significant difference in how we treat functions and methods?

Reply:
I don't think so, I'am going to handle them in the same manner